### ISSUE TITLE
[ops]: ttnn.operations.moreh.clip_grad_norm silently mishandles norm_type = inf / -inf / 0: the p-decimal decomposition is undefined for them and step2 raises the sum to the power 0, while moreh_norm special-cases the same orders

### ISSUE THREAD

**Issue by raunitgrey7** [tenstorrent/tt-metal#55587]
### Component / Area

ops (`ttnn.operations.moreh.clip_grad_norm`, `ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm`)

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.operations.moreh.clip_grad_norm(grads, max_norm, norm_type)` accepts any `norm_type` (the docstring only says "Type of the norm (e.g., 2.0 for L2 norm)"), but the three orders that are not a plain `Sum[|e|^p]^(1/p)` are silently mishandled: `float("inf")`, `float("-inf")` and `0.0`. `inf` is the second most common choice in practice (`torch.nn.utils.clip_grad_norm_(..., norm_type="inf")` / `max_grad_norm` with the infinity norm).

Both device steps decompose `norm_type` into an integer part, a fractional part and a sign:

```cpp
// moreh_clip_grad_norm_step1_program_factory.cpp L19-L25 (same helper duplicated in step2 L18-L24)
std::tuple<uint32_t, float, bool> get_p_decimal_p_is_negative(float ord) {
    auto p = std::floor(ord);
    auto fractional_part = ord - p;
    const bool p_is_negative = p < 0.0f;
    uint32_t integer_part = static_cast<uint32_t>(std::abs(p));
    return std::make_tuple(integer_part, fractional_part, p_is_negative);
}
```

and the kernels then compute `|x|^p * exp(log(|x|) * decimal)` per element (`moreh_common.hpp::power_tile_to_cb`, L887-L960, `power_iterative_tile(dst0, p)` at L908). That decomposition only makes sense for a finite non-zero order:

| `norm_type` | step1: `get_p_decimal_p_is_negative(norm_type)` | step2: `get_p_decimal_p_is_negative(1.0f / norm_type)` (L48) |
|---|---|---|
| `+inf` | `floor(inf) = inf`, `inf - inf = NaN`, `static_cast<uint32_t>(inf)` is undefined (`[conv.fpint]`) | `1/inf = 0` -> `p = 0`, `decimal = 0` -> `x^0 * exp(log(x) * 0) = 1` for every finite `x > 0` |
| `-inf` | same, plus `p_is_negative = true` | `1/-inf = -0.0` -> `p = 0`, `decimal = -0.0`, `p_is_negative = false` -> `1` |
| `0` | `p = 0`, `decimal = 0` -> every element (including the zeros the mask writes into tile padding) contributes `0^0 = 1`, so the per-tensor "sum" is the padded element count | `1/0 = inf` -> undefined integer part, `decimal = NaN` |

Consequences, following the code path element by element:

- `norm_type = inf` / `-inf`: step1 multiplies every element by `exp(log(|x|) * NaN)`, which is NaN by IEEE, and hands the kernel an undefined runtime `p` (the `power_iterative_tile` loop bound). Whatever the SFPU turns that into, step2 then raises it to the power `0`, so the reported `total_norm` is either exactly `1.0` (finite input to step2) or non-finite. It is never `max|g|` / `min|g|`. With `error_if_nonfinite=True` the op can raise the misleading "The total norm of order inf for gradients from `parameters` is non-finite" error for perfectly finite gradients; with the default `False` it silently rescales every gradient by `max_norm / (1.0 + 1e-6)` (L128-L129), i.e. it multiplies all gradients by `max_norm` regardless of their size, or by NaN.
- `norm_type = 0`: the per-tensor count is the padded element count instead of the nonzero count, and step2 gets an undefined integer part and a NaN decimal.

The only guard in the host op is `error_if_nonfinite && std::isnan(norm_type)` (`moreh_clip_grad_norm.cpp` L59); `inf` and `0` pass straight through. The sibling op `ttnn::moreh_norm` special-cases exactly these three orders on the host (`moreh_norm.cpp` L30 and L45, dispatching to the `MINUS_INF` / max / nonzero-count kernels), so the building blocks exist; `clip_grad_norm` just never routes to them. The nightly test only covers `norm_type in [2.0, -0.8, 2.2]` (`test_moreh_clip_grad_norm.py` L28), which is why this went unnoticed.

Concrete example (bfloat16, two gradients, `max_norm = 1.0`, `norm_type = inf`):

```text
g1 = [[ 3.0, -1.0, 0.5, 0.25]]     max|g1| = 3.0
g2 = [[-0.25, 0.125, 0.0, 0.0]]    max|g2| = 0.25

torch  total_norm = 3.0        clip_coef = 1/(3+1e-6)     g1 -> [ 1.0, -0.333, 0.167, 0.083]
ttnn   total_norm = 1.0 (*)    clip_coef = 1/(1+1e-6)     g1 -> [ 3.0, -1.0,   0.5,   0.25 ]  (unclipped)
(*) or NaN / an SFPU-dependent value, see above; in no branch of the code is it 3.0
```

### Expected

`torch.nn.utils.clip_grad_norm_` semantics (`torch.linalg.vector_norm(torch.stack([vector_norm(g, norm_type) for g in grads]), norm_type)`):

- `+inf`: `total_norm = max over tensors of max|g|`
- `-inf`: `total_norm = min over tensors of min|g|`
- `0`: `total_norm = number of tensors whose count of nonzero elements is nonzero`

and the gradients scaled in place by `min(1, max_norm / (total_norm + 1e-6))`, exactly as for `norm_type = 2.0`.

### 1. Steps (exact commands)

Host-only replay of the host-side decomposition and of what the kernels are then asked to compute (no device needed, `pip install torch`):

```python
import math
import torch

def get_p_decimal_p_is_negative(ord):
    # moreh_clip_grad_norm_step1_program_factory.cpp L19-L25 / step2 L18-L24
    p = math.floor(ord) if math.isfinite(ord) else ord
    fractional_part = ord - p
    p_is_negative = p < 0.0
    integer_part = "UB (static_cast<uint32_t> of a non-representable value)" if not math.isfinite(p) else int(abs(p))
    return integer_part, fractional_part, p_is_negative

def step2_scalar(x, p, decimal, p_is_negative):
    # power_tile_to_cb: x^p * exp(log(x) * decimal), IEEE semantics for the transcendental steps
    xpow = x ** p if isinstance(p, int) else float("nan")
    if p_is_negative:
        xpow = 1.0 / xpow
    return xpow * math.exp(math.log(x) * decimal)

for norm_type in (float("inf"), float("-inf"), 0.0):
    print(f"norm_type = {norm_type}")
    print("  step1 (p, decimal, p_is_negative):", get_p_decimal_p_is_negative(norm_type))
    inv = 1.0 / norm_type if norm_type != 0.0 else float("inf")
    p, decimal, neg = get_p_decimal_p_is_negative(inv)
    print("  step2 (p, decimal, p_is_negative):", (p, decimal, neg))
    if isinstance(p, int):
        print("  step2 total_norm for a finite positive Sum[...] of 9.0:", step2_scalar(9.0, p, decimal, neg))

g1 = torch.tensor([[3.0, -1.0, 0.5, 0.25]], dtype=torch.bfloat16)
g2 = torch.tensor([[-0.25, 0.125, 0.0, 0.0]], dtype=torch.bfloat16)
for norm_type in (float("inf"), float("-inf"), 0.0):
    params = []
    for g in (g1, g2):
        p = torch.nn.Parameter(torch.zeros_like(g))
        p.grad = g.clone()
        params.append(p)
    total = torch.nn.utils.clip_grad_norm_(params, 1.0, norm_type)
    print(f"torch norm_type={norm_type}: total_norm={total.item()}  g1 after clip={params[0].grad.tolist()}")
```

Output of the replay:

```text
norm_type = inf
  step1 (p, decimal, p_is_negative): ('UB (static_cast<uint32_t> of a non-representable value)', nan, False)
  step2 (p, decimal, p_is_negative): (0, 0.0, False)
  step2 total_norm for a finite positive Sum[...] of 9.0: 1.0
norm_type = -inf
  step1 (p, decimal, p_is_negative): ('UB (static_cast<uint32_t> of a non-representable value)', nan, True)
  step2 (p, decimal, p_is_negative): (0, -0.0, False)
  step2 total_norm for a finite positive Sum[...] of 9.0: 1.0
norm_type = 0.0
  step1 (p, decimal, p_is_negative): (0, 0.0, False)
  step2 (p, decimal, p_is_negative): ('UB (static_cast<uint32_t> of a non-representable value)', nan, False)
torch norm_type=inf: total_norm=3.0  g1 after clip=[[1.0, -0.333984375, 0.1669921875, 0.08349609375]]
torch norm_type=-inf: total_norm=0.0  g1 after clip=[[3.0, -1.0, 0.5, 0.25]]
torch norm_type=0.0: total_norm=2.0  g1 after clip=[[1.5, -0.5, 0.25, 0.125]]
```

On a card, the same thing as a test:

```bash
pytest tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_clip_grad_norm.py -k special_norm_type
```

(that test is on the branch linked below; on `main` the equivalent is to call `ttnn.operations.moreh.clip_grad_norm([ttnn.from_torch(g1, layout=ttnn.TILE_LAYOUT, device=device), ttnn.from_torch(g2, ...)], 1.0, float("inf"))` and compare against the torch values above).

### 2. Input data / link or description

Inline in the snippet above. Source references are against `main` @ 28238f903b3c35d2f6c91c99d3d249b3db843076:

- `ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp` L59 (only NaN is rejected), L92 / L108 (step1 / step2 dispatch), L128-L129 (`clip_coef`)
- `.../moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp` L19-L25, L63
- `.../moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp` L18-L24, L48 (`1.0f / norm_type`)
- `.../moreh_clip_grad_norm_step1/device/kernels/moreh_clip_grad_norm_step1_kernel.cpp` L14-L15, L106; `.../step2/.../moreh_clip_grad_norm_step2_kernel.cpp` L11-L12, L81
- `ttnn/cpp/ttnn/kernel/compute/moreh_common.hpp` L887-L960 (`power_tile_to_cb`)
- `ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp` L30, L45 (the sibling op's handling of the same orders)
- `tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_clip_grad_norm.py` L28

### 3. Frequency

Always (deterministic; it is the host-side argument decomposition, independent of the data).

### 1. Software Versions

Source-level analysis of `main` @ 28238f903b3c35d2f6c91c99d3d249b3db843076 (2026-09-06). The host-side replay above needs only Python 3 + torch (verified with torch 2.11).

### 2. Hardware Details

Not run on silicon: I do not have a card, so I have not asserted which of the two failure modes (`total_norm == 1.0` vs. a non-finite / SFPU-flushed value) appears on Wormhole or Blackhole. The defect is in the host decomposition and is architecture independent.

### Is this a regression?

Unknown

### Regression Details

The helper and the `1.0f / norm_type` call date back to the original `moreh_clip_grad_norm` implementation; there was never an `inf` / `0` path.

### Logs & Diagnostics

n/a (see replay output above)

### Priority

P2

### Impact

Any training loop using the infinity norm for gradient clipping (`norm_type="inf"`) gets its gradients scaled by `max_norm` (or NaN) instead of being clipped, with no warning. `error_if_nonfinite=True` can turn finite gradients into a hard error.

### Proposed fix

Route `norm_type in {+inf, -inf, 0}` around the step1/step2 kernels and build the total norm from per-tensor `ttnn::moreh_norm` calls (which already implement these orders), combined with the same order the way torch does: `ttnn::maximum` across tensors for `+inf`, `ttnn::minimum` for `-inf`, and the count of tensors with a nonzero per-tensor norm (`ttnn::nez` + `ttnn::add`) for `0`. The optional `total_norm` output tensor is honoured on that path with `ttnn::assign`, and the existing `clip_coef` / step3 code is reused unchanged.

Implemented, with a nightly test for the three orders (`test_moreh_clip_grad_norm_special_norm_type`), on https://github.com/raunitgrey7/tt-metal/tree/fix/moreh-clip-grad-norm-inf-zero-order (diff: https://github.com/tenstorrent/tt-metal/compare/main...raunitgrey7:tt-metal:fix/moreh-clip-grad-norm-inf-zero-order). I would like to be assigned to this issue and will open the PR from that branch.

**Comment by raunitgrey7** [comment]
Thanks for the confirmation. The fix PR is up from that branch: #55593 ([Bug fix] moreh_clip_grad_norm: support norm_type inf, -inf and 0). It routes `norm_type in {+inf, -inf, 0}` around the step1/step2 kernels and builds the total norm from per-tensor `ttnn::moreh_norm` calls combined with the same order (max / min / nonzero-tensor count), honours the optional `total_norm` output, and adds a nightly test for the three orders on aligned and padded shapes.

@jberkowitzTT could this issue be assigned to me, as the reporter with the fix in review?
